File: scripts/dodex_probe/dodex_probe.py
#!/usr/bin/env python3
"""
dodex_probe.py — Python orchestrator for the DODEX Phase 1 read-only probe.

Spawns scripts/dodex_probe/dodex_probe.js as a child process, pipes
JSON-line requests in, reads JSON-line responses out, and prints a short
PASS/FAIL report.

Designed to be run MANUALLY by the operator (or the assistant on demand).
It is NOT wired into the trading-journal service. It signs no transactions.
It writes no DB rows.
"""

from __future__ import annotations

import argparse
import json
import os
import subprocess
import sys
import time
from pathlib import Path

HERE = Path(__file__).resolve().parent
ENDPOINTS = json.loads((HERE / "endpoints.json").read_text())


# ── Sidecar ──────────────────────────────────────────────────────────────────

class Sidecar:
    def __init__(self) -> None:
        sidecar_js = HERE / "dodex_probe.js"
        if not (HERE / "node_modules").exists():
            print("⚠  node_modules/ missing — run `npm install` in this directory first")
            print(f"   cd {HERE} && npm install")
            sys.exit(2)
        self.proc = subprocess.Popen(
            ["node", str(sidecar_js)],
            cwd=HERE,
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            bufsize=1,
        )
        self._id = 0

    def call(self, op: str, args: dict | None = None, timeout_s: int = 25) -> dict:
        self._id += 1
        msg = {"id": self._id, "op": op, "args": args or {}}
        line = json.dumps(msg) + "\n"
        assert self.proc.stdin is not None and self.proc.stdout is not None
        self.proc.stdin.write(line)
        self.proc.stdin.flush()

        t0 = time.time()
        while True:
            if time.time() - t0 > timeout_s:
                return {"ok": False, "error": "timeout"}
            resp_line = self.proc.stdout.readline()
            if not resp_line:
                # sidecar died — capture stderr
                stderr = self.proc.stderr.read() if self.proc.stderr else ""
                return {"ok": False, "error": "sidecar died", "stderr": stderr.strip()}
            try:
                return json.loads(resp_line)
            except json.JSONDecodeError:
                # garbage from a console.log in node — skip and keep reading
                continue

    def close(self) -> None:
        try:
            if self.proc.stdin:
                self.proc.stdin.close()
            self.proc.wait(timeout=5)
        except Exception:  # noqa: BLE001
            try:
                self.proc.kill()
            except Exception:  # noqa: BLE001
                pass


# ── Steps ────────────────────────────────────────────────────────────────────

def step(name: str, fn) -> bool:
    print(f"  → {name} ...", end=" ", flush=True)
    t0 = time.time()
    try:
        ok, detail = fn()
    except Exception as e:  # noqa: BLE001
        print(f"FAIL ({time.time()-t0:.2f}s)")
        print(f"      exception: {type(e).__name__}: {e}")
        return False
    elapsed = time.time() - t0
    if ok:
        print(f"PASS ({elapsed:.2f}s)")
        if detail:
            print(f"      {detail}")
    else:
        print(f"FAIL ({elapsed:.2f}s)")
        print(f"      {detail}")
    return ok


def run(network: str = "mainnet") -> int:
    net = ENDPOINTS.get(network)
    if not net:
        print(f"✗ unknown network '{network}', use one of: {list(ENDPOINTS)}")
        return 2

    print(f"\nDODEX Phase 1 probe — {net['name']}")
    print(f"endpoint root: {net['root']}\n")

    sc = Sidecar()
    passes = 0
    fails = 0

    # 1. sidecar sanity check
    def ping():
        r = sc.call("ping")
        return (r.get("ok"), f"sdk_version_hint={r.get('result',{}).get('sdk_version_hint')}")
    if step("sidecar ping (Node + TVM-SDK loaded)", ping):
        passes += 1
    else:
        fails += 1
        sc.close()
        print(f"\n  aborting — sidecar not healthy")
        return 1

    # 2. for each GraphQL candidate, try node_info
    candidates = net.get("graphql_candidates") or [net.get("graphql")]
    candidates = [c for c in candidates if c]
    working_endpoint = None
    for endpoint in candidates:
        def query(ep=endpoint):
            r = sc.call("node_info", {"endpoint": ep})
            if r.get("ok"):
                return (True, f"{ep} ← OK · {json.dumps(r.get('result'))[:160]}")
            return (False, f"{ep} ← {r.get('error','?')[:120]}")
        if step(f"GraphQL discovery: {endpoint}", query):
            working_endpoint = endpoint
            passes += 1
            break
        fails += 1

    if not working_endpoint:
        print("\n  ✗ no GraphQL endpoint responded — Phase 1 cannot proceed until")
        print("    we find the live endpoint or DODEX docs publish one.")
        sc.close()
        return 1

    # 3. block height check
    def block():
        r = sc.call("block_height", {"endpoint": working_endpoint})
        if r.get("ok") and r.get("result"):
            top = (r["result"] or [{}])[0]
            return (True, f"top block seq_no={top.get('seq_no')} gen_utime={top.get('gen_utime')}")
        return (False, str(r.get("error") or "no result"))
    if step("query top block", block): passes += 1
    else: fails += 1

    # ── Summary ──────────────────────────────────────────────────────────────
    sc.close()
    print(f"\n{passes} passed · {fails} failed")
    print(f"working endpoint: {working_endpoint}")
    return 0 if fails == 0 else 1


if __name__ == "__main__":
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--network", default="mainnet",
                    choices=list(ENDPOINTS.keys()) if "_comment" not in ENDPOINTS else None,
                    help="endpoint key in endpoints.json")
    args = ap.parse_args()
    sys.exit(run(args.network))

File: scripts/dodex_probe/dodex_probe.js
/**
 * dodex_probe.js — Node sidecar for the DODEX Phase 1 read-only probe.
 *
 * Talks to Acki Nacki via the official TVM-SDK (@tvmsdk/lib-node).
 * Communicates with the Python orchestrator via newline-delimited JSON
 * on stdin / stdout. NEVER signs or submits a transaction.
 *
 * Protocol:
 *   stdin  ← {"id": <int>, "op": <op>, "args": {...}}
 *   stdout → {"id": <int>, "ok": true/false, "result": ..., "error": ...}
 *
 * Supported ops (all read-only):
 *   "ping"          — sanity check, returns {pong: true, sdk_version: <str>}
 *   "node_info"     — query node version / lastBlockTime via GraphQL
 *   "block_height"  — current last seen block
 *   "account_state" — fetch a public address's state (balance, code hash)
 *   "run_local"     — emulate a get-method call on a contract (off-chain)
 *
 * To add a write capability later, declare it explicitly here AND require
 * a separate "write_unlock" op so it's hard to ship by accident.
 */
'use strict';

const readline = require('node:readline');

// Lazy-load the SDK so the script can fail with a friendly message if
// `npm install` hasn't been run yet.
let TonClient, libNode;
try {
    ({ TonClient } = require('@tvmsdk/core'));
    libNode = require('@tvmsdk/lib-node');
} catch (e) {
    process.stdout.write(JSON.stringify({
        id: 0, ok: false,
        error: 'tvm-sdk not installed — run `npm install` in scripts/dodex_probe/',
        detail: String(e),
    }) + '\n');
    process.exit(1);
}

// SDK bootstrap
TonClient.useBinaryLibrary(libNode.libNode);

// One client per endpoint, cached
const clients = new Map();
function getClient(endpoint) {
    if (clients.has(endpoint)) return clients.get(endpoint);
    const c = new TonClient({ network: { endpoints: [endpoint] } });
    clients.set(endpoint, c);
    return c;
}

// ── Handlers ────────────────────────────────────────────────────────────────

async function handle(msg) {
    const { id, op, args = {} } = msg;
    try {
        switch (op) {
            case 'ping':
                return { id, ok: true, result: { pong: true, sdk_version_hint: TonClient.version || 'unknown' } };

            case 'node_info': {
                const client = getClient(args.endpoint);
                const versionRes = await client.net.query({
                    query: '{ info { version time lastBlockTime } }',
                });
                return { id, ok: true, result: versionRes };
            }

            case 'block_height': {
                const client = getClient(args.endpoint);
                const res = await client.net.query_collection({
                    collection: 'blocks',
                    filter: { workchain_id: { eq: 0 } },
                    order: [{ path: 'seq_no', direction: 'DESC' }],
                    limit: 1,
                    result: 'id seq_no gen_utime',
                });
                return { id, ok: true, result: res.result };
            }

            case 'account_state': {
                const client = getClient(args.endpoint);
                if (!args.address) throw new Error('account_state requires `address`');
                const res = await client.net.query_collection({
                    collection: 'accounts',
                    filter: { id: { eq: args.address } },
                    result: 'id acc_type balance code_hash last_paid',
                });
                return { id, ok: true, result: res.result };
            }

            case 'run_local': {
                const client = getClient(args.endpoint);
                if (!args.address) throw new Error('run_local requires `address`');
                if (!args.abi) throw new Error('run_local requires `abi`');
                if (!args.method) throw new Error('run_local requires `method`');
                const acc = await client.net.query_collection({
                    collection: 'accounts',
                    filter: { id: { eq: args.address } },
                    result: 'boc',
                });
                if (!acc.result || acc.result.length === 0) {
                    return { id, ok: false, error: 'account not found' };
                }
                const res = await client.tvm.run_tvm({
                    account: acc.result[0].boc,
                    message: {
                        abi: { type: 'Contract', value: args.abi },
                        address: args.address,
                        call_set: { function_name: args.method, input: args.input || {} },
                        signer: { type: 'None' },
                        is_internal: false,
                    },
                });
                return { id, ok: true, result: res.decoded };
            }

            default:
                return { id, ok: false, error: `unknown op ${op}` };
        }
    } catch (e) {
        return { id, ok: false, error: String(e && e.message || e) };
    }
}

// ── Main I/O loop ───────────────────────────────────────────────────────────

const rl = readline.createInterface({ input: process.stdin });

rl.on('line', async (line) => {
    line = line.trim();
    if (!line) return;
    let msg;
    try {
        msg = JSON.parse(line);
    } catch (e) {
        process.stdout.write(JSON.stringify({ id: 0, ok: false, error: 'invalid json on stdin', detail: String(e) }) + '\n');
        return;
    }
    const reply = await handle(msg);
    process.stdout.write(JSON.stringify(reply) + '\n');
});

rl.on('close', () => process.exit(0));
